File: 230 Kth Smallest Element in a BST.c

"""
Created on Tues Nov. 10 19:29 2020
Old version and rewrite here
@author: Chiwen Chen
"""



/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */



int kthSmallest(struct TreeNode* root, int k){
    int count = k;
    int eletval = 0;
    
    flatten(&count, &eletval, root);
    return eletval;
}

void flatten(int *count, int *eletval, struct TreeNode *node){
    if (node == NULL) 
    { 
        return; 
    } 
    flatten(count, eletval, node->left);
    
    *count -= 1; 
    
    if (*count == 0) {
        *eletval = node->val; return; 
    } 
    flatten(count, eletval, node->right); 
}

"""
Runtime: 8 ms, faster than 96.69% of C online submissions for Kth Smallest Element in a BST.
Memory Usage: 10.2 MB, less than 16.02% of C online submissions for Kth Smallest Element in a BST.
"""

File: 230 Kth Smallest Element in a BST..py

"""
Created on Tues Nov. 10 21:49 2020
Old version and rewrite here
@author: Chiwen Chen
"""


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

def dfs(self, root, ansval):
    if root.left != None:
        dfs(self, root.left, ansval)
    ansval.append(root.val)
    if root.right != None:
        dfs(self, root.right, ansval)
class Solution:
    def kthSmallest(self, root: TreeNode, k: int) -> int:
        ansval=[]
        dfs(self, root, ansval)
        return ansval[k-1]
    
"""Runtime: 8 ms, faster than 96.69% of C online submissions for Kth Smallest Element in a BST.
Memory Usage: 10.2 MB, less than 16.02% of C online submissions for Kth Smallest Element in a BST."""